Refactor/cooker app response

joelnguemeta opened the pull request:
### **Changements Principaux**

- **Standardisation des Réponses API** : Mise en place d'une enveloppe JSON unifiée pour toutes les réponses , incluant systématiquement un indicateur de succès, les données métiers, un message et un horodatage.

- Isolation renforcée des données pour garantir que chaque cuisinier ne voit que ses propres informations.

**Infrastructure et Sécurité :**
- Gestion dynamique des clés de sécurité RSA adaptée aux environnements locaux et de développement.

- Simulation automatique des validations OTP en mode développement pour fluidifier le travail des équipes(j'ai juste supposé que OTP en local est 000000 pour aller plus vite).

-   les les cles RSA avait un soucis de formatage, j ai fomaté la clé seulement lorsque nous somme en local pour ne pas touché ce que tu as mis en place

- Standardisation du format des jetons d'authentification pour une cohérence globale.


Réussite confirmée de l'intégralité de la suite de tests du module cuisinier (90 tests validés).

Commit:
refactor(cooker_app): -patch send_opt on dev or local mode
-patch verify opt on local mode

Commit:
refactor(settings): - format rsa key on local or dev

Commit:
Refactor CookerApp response format and fix associated tests

Commit:
Update read API tests

Commit:
refactor: standardized token views response and fixed tests

Commit:
fix(test): update dashboard view tests to standardized response response format

Commit:
update DishView to use standard response and fix empty list bug

Commit:
refactor DrinkView to use standard response and update tests

Commit:
fix(cooker-app): correct order filtering and align serializer with tests

- Fixed CookerOrderView to filter by cooker__id instead of customer__id.
- Updated CookerOrderGETSerializer with nested cooker information.
- Updated CookerOrderCustomerGETSerializer to include stripe_id and match expectations.
- Adjusted test data in test_api.py to correctly reflect orders belonging to Cooker 1.

Commit:
Standardize API responses for customer_app and maintain token standardization

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -41,6 +42,43 @@ class Meta:
+            formatted_phone = phonenumbers.format_number(
+                parsed_phone, phonenumbers.PhoneNumberFormat.NATIONAL
+            ).replace(" ", "")
+        except Exception:
```
Je pense qu'on peut être plus précis sur le type d'exception quand on utilise la lib phonumbers, voir la fonction format_phone dans le fichier `utils/common.py`

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -41,6 +42,43 @@ class Meta:
+        except Exception:
+            formatted_phone = data["phone"]
+
+        return {
```
Je n'ai rien contre le fait de le faire dans la couche serializer au lieu du renderer mais il ne faut donc pas oublier de supprimer le code correspondant dans le renderer (s'il n'est plus utilisé)

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -19,7 +19,8 @@
 )
-from utils.common import compute_order_items_total_amount, format_phone
+from utils.common import compute_order_items_total_amount, format_phone, get_pre_signed_url
+import phonenumbers
```
Quand je vois l'ordre des imports, est-ce que tu as bien installé pre-commit ? c'est un outil qui est déjà dans le projet et qui permet à chaque commit  d'appliquer un certain nombre de règles sur le code pour que tous les devs aient le même style de code.

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -41,6 +42,43 @@ class Meta:
+                },
+            },
+            "address_section": {
+                "title": "address",
```
Quitte à refaire les formats de sortie, voir avec le front si la clef "title" est toujours d'actualité et/ou nécessaire. Pourquoi ? je pense quand on a une clef qui s'appelle `"personal_infos_section"` j'ai l'impression qu'avoir encore une clef `"title": "personal_infos"` me semble un peu redondant.

joelnguemeta replied:
J"ai pris le temps de supprimer les "title" comme convenu et aussi "data' dans chaque section. Plus que "data" est la seule clé, je n'ai pas trop vu l importence d avoir une cle impbriqué, meme si c est un O(1) opreation

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -41,6 +42,43 @@ class Meta:
+
+        return {
+            "personal_infos_section": {
+                "title": "personal_infos",
```
Voir mon prochain commentaire pour la clef `title`

joelnguemeta replied:
Il sera a jour lors du prochain push

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -165,7 +203,18 @@ class Meta:
             "id",
             "firstname",
             "lastname",
-            "is_deleted",
```
Juste pour être sûr: pourquoi cette suppression ?

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -174,6 +223,7 @@ class CookerOrderGETSerializer(ModelSerializer):
     dishes_items = OrderDishItemGETSerializer(many=True)
     drinks_items = OrderDrinkItemGETSerializer(many=True)
     customer = CookerOrderCustomerGETSerializer()
+    cooker = CookerOrderCookerGETSerializer()
```
Sauf erreur de ma part tu ajoutes un nouveau serializer CookerOrderCookerGETSerializer qui pointe sur CookerModel mais dans le même fichier il y'a un `CookerSerializer` et `CookerGETSerializer` qui utilisent également `CookerModel`. Dans ton `CookerOrderCookerGETSerializer` tu inclus des champs déjà présents dans `CookerGETSerializer`.

Alors peut-être que j'ai raté un truc donc n'hésite pas à m'éclairer

toutiii replied:
Si la logique est de ne pas toucher aux renderers pour l'instant, c'est plus cohérent de tout faire au même endroit. Autant déporter la logique des renderers vers les serializers, supprimer les renderers et mettre à jour le code aux endroits appropriés (là où on utilisait les renderers avant)

toutiii replied:
Au vu de la taille de la PR et du fait qu'elle est ouverte depuis quelques jours déjà, cette partie sera traitée dans un ticket à part pour fluidifier le dev

toutiii commented on reats/source/settings.py:
```diff
@@ -105,8 +105,23 @@
     print("Secrets loaded successfully")
-    RSA_PRIVATE_KEY_PATH = secret_data["RSA_PRIVATE_KEY_PATH"]
-    RSA_PUBLIC_KEY_PATH = secret_data["RSA_PUBLIC_KEY_PATH"]
+    if os.environ["ENV"] == "local":
```
Juste pour être sûr: l'idée c'est de créer un "vrai" token quand on dev en local avec le front démarré (toujours en local) qui envoie des requêtes sur notre container qui tourne en local ?

toutiii commented on reats/tests/cooker_app/cookers/update/test_api.py:
```diff
@@ -437,8 +437,8 @@ def test_response(
-        assert response.json().get("error_code") == "token_not_valid"
-
+        assert response.json().get("error").get("code") == "token_not_valid"
+    
```
Trailing whitespace  => je pense qu'effectivement pre-commit n'est pas installé car c'est typiquement le genre de choses qu'il va cleaner pour toi

toutiii commented on reats/utils/custom_api_reponse.py:
```diff
@@ -0,0 +1,119 @@
+                status_code=status.HTTP_401_UNAUTHORIZED
+            )
+
+        if response is None:
```
Je pense qu'on peut merger ces deux ifs avec un `and`

toutiii commented on reats/utils/custom_api_reponse.py:
```diff
@@ -0,0 +1,119 @@
+
+        if response is not None:
+
+            if isinstance(response.data, dict) and "success" in response.data:
```
Question

Est ce que le `if response is not None:` est vraiment nécessaire ?

Vu que tu testes en dessous `isinstance(response.data, dict)`. Si `response` est None `isinstance(response.data, dict)` retournera de toute façon False je pense ?

En fait il y'a beaucoup de `if` dans la fonction et en général j'essaie de les simplifier pour faciliter la lecture

toutiii commented on reats/utils/custom_api_reponse.py:
```diff
@@ -0,0 +1,119 @@
+                return response
+
+            message = "Operation failed"
+            code = "API_ERROR"
```
Ces deux strings hardcodés pourraient être mis dans un `__init__` ?

toutiii commented on reats/utils/custom_api_reponse.py:
```diff
@@ -0,0 +1,119 @@
+            code = "API_ERROR"
+            details = response.data
+
+            if isinstance(details, dict):
```
Pour faciliter la lecture, je pense qu'on peut mettre ce bloc

```
if isinstance(details, dict):
                if "detail" in details:
                    if getattr(details["detail"], "code", None):
                        code = details["detail"].code
                    message = str(details["detail"])
                    if message == "Given token not valid for any token type":
                        message = "Token is invalid or expired"
```

dans une méthode genre `_get_message` (ou tout autre non plus pertinent)

toutiii commented on reats/utils/custom_api_reponse.py:
```diff
@@ -0,0 +1,119 @@
+            "success": True,
+            "data": data if data is not None else {},
+            "message": message,
+            "timestamp": datetime.now(timezone.utc).isoformat()
```
C'est une demande du front d'avoir le timestamp de la génération de la réponse ? parce que c'est quand même rare qu'on en ait besoin sauf peut-être quand on log des choses...

toutiii commented on reats/utils/custom_api_reponse.py:
```diff
@@ -0,0 +1,119 @@
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework_simplejwt.exceptions import InvalidToken, TokenError
+import sys
```
import non utilisé ? Si oui il faut vraiment run un pre-commit parce ta pipeline échouera si le pre-commit n'est pas run quand tu commit en local.

toutiii commented on reats/utils/common.py:
```diff
@@ -138,6 +146,11 @@ def send_otp(phone: str) -> Union[dict, None]:
 
 def is_otp_valid(data: dict) -> bool:
+    
+    if os.environ["ENV"] == "local" and data["otp"] == "000000":
```
Alors je comprends l'intention mais je ne suis pas forcément pour ce type de code. Je m'explique

1. Même si le risque est minime, çà veut dire qu'il n'est pas impossible de remplacer "local" par "prod" et avoir un OTP qui passe tout le temps. Ce n'est pas censé arriver car on essaie de faire gaffe à ce qu'on merge mais une review rapide un vendredi soir par un reviewer peu attentif peut mener à un déploiement en prod de cette faille

2. On laisse clairement une backdoor dans le code et çà peut se retourner contre nous. (Attention ce repo est public. Tout le monde ne peut pas y contribuer mais tout le monde peut le trouver et donc le lire)

3. C'est pareil je ne suis pas favorable au code qui comporte par exemple des `if env=="test"`. Cela signifie qu'il y'a un problème d'architecture et/ou de conception. Pour moi le code qu'on teste est le même peu importe l'env, ce sont les `input` qui changent (fichiers .env, fixtures, etc)

Après avoir dit çà quelle est la solution ?

Une solution possible qui me paraît acceptable en local pour la partie auth/OTP en local serait la suivante.

En soit en local on dev des features donc on bypass dans 95% des cas le processus d'auth (sinon même développer devient pénible et les OTP ce n'est pas gratuit). Donc un dev pourrait modifier le code et faire

```
 @action(methods=["post"], detail=False, url_path="otp-verify")
    def otp_verify(self, request) -> Response:
        return Response(status=status.HTTP_200_OK)  # LIGNE AJOUTÉE de bypass
        result = is_otp_valid(request.data)
```

Donc en local l'OTP est systématiquement bypass. Pourquoi je pense que c'est une solution viable:

1. C'est une action manuelle et çà fait sens dans la mesure où on désactive une feature critique.
2. L'intention est claire => on veut bypass l'OTP
3. En cas d'oubli du code en local ==> certains tests vont échouer et c'est très bien, les tests nous servent de garde-fous
4. Si jamais le code est commité par erreur  => pareil la pipeline va casser et des erreurs sur les tests unitaires nous seront remontés.


Dans ton exemple du if `env=="local"` si `"local" `est remplacé par `"prod",` pas sûr que la pipeline nous protège de manière aussi explicite.

Pour info à terme l'auth va être totalement sortie du code et gérée par une lambdaAuthorizer côté AWS donc ces contraintes en local de désactivation de l'auth/OTP vont disparaître.

toutiii commented on reats/utils/custom_api_reponse.py:
```diff
@@ -0,0 +1,119 @@
+                "message": message,
+                "details": details or {}
+            },
+            "timestamp": datetime.now(timezone.utc).isoformat()
```
même question pour le timestamp ici

toutiii commented on reats/utils/common.py:
```diff
@@ -114,6 +114,14 @@ def generate_ref_id(phone: str):
 
 
 def send_otp(phone: str) -> Union[dict, None]:
+    if os.environ["ENV"] == "local":
```
Voir mon long commentaire plus bas sur les `if "env=="local"`

toutiii commented on reats/customer_app/views.py:
```diff
@@ -291,11 +292,8 @@ def destroy(self, request, *args, **kwargs) -> Response:
-                "status_code": status.HTTP_200_OK,
-            }
+        return self.success(
+            message="Address deleted successfully"
```
Est ce que çà ne vaut pas le coup de faire un Enum pour çà dans common/ (par exemple) pour centraliser les messages, les codes de retours.

Par exemple si on veut changer "Operation Sucessful" par autre chose il faudra quand même modifier plusieurs endroits dans le code et ces enums pourraient être utilisés dans les tests unitaires également pour éviter du texte en dur

toutiii commented on reats/cooker_app/serializers.py:
```diff
@@ -165,7 +203,18 @@ class Meta:
             "firstname",
             "lastname",
-            "is_deleted",
+            "stripe_id",
```
Sauf erreur de ma part, le cooker n'a pas de `stripe_id` à ce jour. Seul `CustomerModel` en possède

joelnguemeta replied:
C'est le models customer qui entre en jeu. J avais pensé que le cusinier devait en avoir besoin pour remboursser le client, mais je me rend compte que les infos tels que le nom, prenom sont deja suiffisant et en plus c est proccess qui sera en arrière plan.

Merci pour la remarque

toutiii commented on reats/tests/conftest.py:
```diff
@@ -17,6 +17,12 @@
 
 
+@pytest.fixture(autouse=True)
+def mock_env_vars():
```
Juste pour être sûr de comprendre: quelle est l'utilité de cette fixture ?

joelnguemeta replied:
Plus que j avais mis les "if os.getenv == local"un peu de partout( cle RSA et OTP), c etait pour mocker aussi le environement de travail.

Mais , il a ete retiré aujourd'hui

Commit:
refactor(cooker_app): ajout de NumberParseException dans CookerGerSerializer

Commit:
refactor(cooker_app): suppression de CookerCustomRendererWithData dans renderes

Commit:
refactor(cooker_app): Suppressions des messages en dure dans cooker_app et validation des tests

- Ajout de la classe ErrorCodeEnum pour les code d erreur
- Ajout de ErrorMessageEnum pour les messages d erreur
- Ajout de SuccessMessageEnum pour les messages de success
- Remplacement de toutes les occurences des messages en dure dans reats/cooker_app/views.py
- Meme chose pour les tests de cooker_app

Commit:
refactor(cooker_app): suppression de strip_id

Commit:
refactor(cooker_app): suppression des timestamps

Commit:
refactor(cooker_app): Merge if statement in StandardizedResponseMixin

Commit:
refactor(cooker_app): suppression de la condition if dans send_otp

Commit:
refactor: - Correction des test, suppression de la fonction fromat_pem_key, ajout de la logique de gestion d erruer pour eviter les reponse Null

Commit:
refactor: ajout de la classe Exception handler pour plus de maintenance

toutiii commented on .pre-commit-config.yaml:
```diff
@@ -35,4 +35,4 @@ repos:
           - djangorestframework==3.14.0
           - python-dotenv==1.0.0
-          - psycopg2==2.9.6
+          - psycopg2-binary==2.9.11
```
Pourquoi cette montée de version ? il y avait un souci avec la 2.9.6 ?

joelnguemeta replied:
J ai debuté avec docker-compose, mais il falait chaque fois, relancer le container a chaque modification alors que par monent, j'ai juste besoin d'un reload vite fais, j'ai preferer mettre en place en venv en local chez  moi et demarer le projet. psycopg2 ne s'intalle pas sur mon OS(linux), j ai ajouté la version binaire qui est tres supporté par linux

toutiii commented on reats/custom_renderers/renderers.py:
```diff
@@ -12,75 +12,75 @@
-
-        logger.info(response)
-        return super().render(response)
+# class CookerCustomRendererWithData(JSONRenderer):
```
Ne pas garder du code commenté (surtout si les tests passent) ==> on peut le supprimer sans crainte

joelnguemeta replied:
C est fait a ce niveau et sera a jour lors du prochain push

toutel committed:
Fixing docker compose warning

toutel committed:
Avoid unexpected update for pg version release when building new image

toutel committed:
Fixing token issues in local env

toutel committed:
Ruff auto fix

toutel committed:
Updating dot env files with right RSA keys name in SM

Commit:
refactor(cooker_app): - suppression des cle data et title,

toutel committed:
Fixing unit tests

toutel committed:
Fixing pr ci file

toutel committed:
Fixing pr ci file
